### PR TITLE
Enhanced features

### PR DESCRIPTION
**Pull request by phomer**
Filled in the features.

Some deficiencies:
- Not sure the hostname is handled properly
- The streaming is clunky
- Needs nicer error messages
- Needs better comments
- Needs better testing
- gorilla/mux seems to collide with signals and fsnotify, not sure why.

### CODE REVIEW

File: accounts/accounts.go [hunk 1]
```diff
@@ -46,41 +47,67 @@ func NewAccount(username string, hostname string) *Account {
...
 }
 
-// TODO: Not threadsafe, or persistent
-func (account *Account) IncrementId() {
-	// TODO: Needs to be done correctly
-	account.NextId += 1
-}
+var current *Authentication
 
 func NewAuthentication() *Authentication {
```
Comment (knisbet) [inline accounts/accounts.go:56]: nit: Naming, to me NewAuthentication implies that this is a builder (in java terms a constructor) for an object typ, but the latest changes appear to return more like a Singleton. The way the golang standard library appears to deal with this, using DNS as an example, is by having DefaultResolver. So in this case the function could be called something more like DefaultAuthentication to return the default implementation that is locally relevant.

File: accounts/auth.go [hunk 2]
```diff
@@ -32,32 +34,62 @@ func Username() string {
...
 
 	tokenString, err := token.SignedString(SecretKey())
 	if err != nil {
-		fmt.Println("Failed to Sign Token", err, token)
+		fmt.Println("Failed to Sign Token ", err, token)
 		panic("Goodbye")
 	}
 
+	// Test it
+	_, err = jwt.ParseWithClaims(tokenString, claims, GetKey)
```
Comment (knisbet) [inline accounts/auth.go:61]: Sorry, I'm not sure I understand why you would test parsing the claim here? Why not setup some unit tests if the token generation is a concern?

File: accounts/config.go [hunk 3]
```diff
@@ -59,13 +81,19 @@ func (config *ClientConfig) SaveConfig() {
...
+		if err != nil {
+			return "", err
+		}
+
+		// TODO: Should realy ask the user which one they want, but for
+		// now lets just take the first file that is found.
+		if len(matches) > 0 {
+			return matches[0], nil
+		}
 	}
```
Comment (pierrebeaucamp) [inline accounts/config.go:97]: This still doesn't seem to check the different paths. Is that still a `TODO` item?

File: comm/handlers.go [hunk 4]
```diff
@@ -14,65 +18,296 @@ import (
...
+		} else {
+			WriteResponse(response, "Status: ", "Finished")
+		}
+	} else {
+		WriteResponse(response, "Status: ", "Pending")
+	}
+}
+
+// Handle logging and sending back and error
+func HandleError(response http.ResponseWriter, err interface{}) {
```
Comment (pierrebeaucamp) [inline comm/handlers.go:185]: Why is `err` an interface here? Seems like we expect an `error` and the casting below could lead to panics

File: comm/handlers.go [hunk 5]
```diff
@@ -14,65 +18,296 @@ import (
...
+		job.Offset, err = job.File.Seek(job.Offset, 0)
+		if err != nil {
+			HandleError(response, err)
+		}
+	}
+
+	// Keep trying until the job stops, it runs out of file, or the pipe closes.
+	finished := false
+	for {
+		time.Sleep(1000 * time.Millisecond)
```
Comment (pierrebeaucamp) [inline comm/handlers.go:276]: This seems very similar to a `time.Ticker`. Is there any specific reason for rolling a similar thing with sleep statements?

File: accounts/accounts.go [hunk 6]
```diff
@@ -46,41 +47,67 @@ func NewAccount(username string, hostname string) *Account {
...
+		Token:     token,
 	}
 }
 
-// TODO: Not threadsafe, or persistent
-func (account *Account) IncrementId() {
-	// TODO: Needs to be done correctly
-	account.NextId += 1
-}
+var current *Authentication
```
Comment (pierrebeaucamp) [inline accounts/accounts.go:54]: Are there any trade-offs by making this global? Or is this just a convenience for the time being?

File: accounts/accounts.go [hunk 7]
```diff
@@ -46,41 +47,67 @@ func NewAccount(username string, hostname string) *Account {
...
 }
 
-func (auth *Authentication) Find(username string) *Account {
+func (auth *Authentication) Reload() {
+	auth.mux.Lock()
+	auth.file_lock()
+
+	auth.load()
+
+	auth.file_unlock()
```
Comment (knisbet) [inline accounts/accounts.go:103]: tip: use defer file_unlock when taking the lock, this way regardless of the way the function returns the unlock will get called.

```
auth.file_lock()
defer auth.file_unlock()
```

It just makes it alot harder that if this function get's changed in the future for multiple returns, that the unlock is consistently called.

File: comm/handlers.go [hunk 8]
```diff
@@ -14,65 +18,296 @@ import (
...
 		}
 	}()
 
 	account, err := ValidateRequest(response, request)
 	if err != nil {
-		log.Fatal("Validate", err)
+		HandleError(response, err)
 	}
 
+	fmt.Println("Immediate Command")
```
Comment (knisbet) [inline comm/handlers.go:34]: include some context when logging. A log that just says Immediate Command doesn't tend to be very usefull to tracing what is happening or going wrong.

File: comm/handlers.go (hunk 4 above)
Comment (phomer) [reply comm/handlers.go:185]: It's because I'm using it in recover() as well, which seemed to return its param as an interface{}.

File: comm/handlers.go (hunk 5 above)
Comment (knisbet) [inline comm/handlers.go:276]: I don't believe this sleep is needed, I don't see what sleeping here is valuable for??

File: comm/handlers.go [hunk 9]
```diff
@@ -14,65 +18,296 @@ import (
...
+		}
+	}
+
+	// Keep trying until the job stops, it runs out of file, or the pipe closes.
+	finished := false
+	for {
+		time.Sleep(1000 * time.Millisecond)
+		buffer := make([]byte, 1024)
+
+		count, err := job.File.Read(buffer)
```
Comment (knisbet) [inline comm/handlers.go:279]: I don't see anywhere that read deadlines are set, so I believe this Read call should be blocking on data being available. It's not necessary to throttle the reads below with sleep calls, since this won't busy loop.

File: comm/handlers.go (hunk 5 above)
Comment (phomer) [reply comm/handlers.go:276]: No, it was just easier for me to think about and I was really short on time. That Sleep probably shouldn't be there either, but I discovered at the last moment that there was a delay between the process getting reaped and its buffers getting synced to the filesystem (which was kinda a surprise, but it's quite been a while since I've written this type of code).

File: comm/server.go [hunk 10]
```diff
@@ -99,26 +116,79 @@ func HandleSighup() {
...
+	// TODO: SIGHUP seems to hang the web server, not sure if this is
 	// a reasonable way to restart it, or it's just leaking ...
-	err := server.web.ListenAndServe()
-	if err != nil {
-		fmt.Println("FATAL: Web Server Error: ", err)
-		return
-	}
+	// Probably need a better way to tell the server to reload
+	fmt.Println("Restarting Web Server after Signal")
+	TryWebServerRestart()
```
Comment (knisbet) [inline comm/server.go:128]: I really wouldn't worry about any of this signal based reloading, it feels really like scope creep on what's just meant to be a prototype. There is nothing wrong with just having to restart a process wen something external changes.

File: accounts/accounts.go (hunk 6 above)
Comment (phomer) [reply accounts/accounts.go:54]: I had started with encapsulating everything into an overall 'server' context, that was to be passed down to avoid having any globals. However, the handlers for net/http needed access, so I had switched to that horrible Global() call. But then, just to get it done, I basically made all of the parent data structs as Singletons (even though I really hate that pattern). So, really just convenience, and now somewhat inconsistent :-(

File: comm/handlers.go (hunk 4 above)
Comment (pierrebeaucamp) [reply comm/handlers.go:185]: Hm, to be honest, this is a bit out of my ballpark because I rarely use `recover`. My understanding is that `recover` returns whatever is given to `panic`, hence the empty interface. But this doesn't necessarily need to be an error.

Particularly, I think that this `HandleError` method will panic if the recovered value itself is a string. As an example: https://play.golang.org/p/37jhPYnN-z_X

I would prefer to change `HandleError` to accept an `error` type instead to make this safe and have the function that `recover`s from a panic deal with casting the value into an acceptable error type.

File: comm/handlers.go (hunk 5 above)
Comment (pierrebeaucamp) [reply comm/handlers.go:276]: "Easier to implement" is a totally valid reason for me in the context of this challenge. Just wanted to understand what lead to the current code.

File: accounts/accounts.go [hunk 11]
```diff
@@ -46,41 +47,67 @@ func NewAccount(username string, hostname string) *Account {
...
+	auth.load() // Could have changed since last loaded
 	auth.Map.Accounts[account.Username] = account
+	auth.store()
 
+	auth.file_unlock()
 	auth.mux.Unlock()
 }
 
-func (auth *Authentication) Find(username string) *Account {
+func (auth *Authentication) Reload() {
```
Comment (r0mant) [inline accounts/accounts.go:97]: This whole Authentication struct looks pretty complicated and error prone with multiple locks, reloads, etc. Why is it necessary? Is it possible to simplify the authentication layer to avoid using files, etc? Like mentioned before, it is okay (and even preferable for the sake of the challenge) to keep the authentication layer as simple as possible and accept some pre-generated credentials on the CLI.

File: accounts/accounts.go [hunk 12]
```diff
@@ -90,24 +117,27 @@ func (auth *Authentication) Find(username string) *Account {
...
 }
 
+// TODO: Quick fix to expose the file name
+func (auth *Authentication) GetFilepath() string {
+	return auth.db.GetFilepath()
+}
+
 // Lock the underlying accounts file
-func (auth *Authentication) Lock() {
+func (auth *Authentication) file_lock() {
```
Comment (r0mant) [inline accounts/accounts.go:126]: We follow https://github.com/golang/go/wiki/CodeReviewComments, so in Go style all methods and variables are named using camelCase.

File: accounts/auth.go [hunk 13]
```diff
@@ -32,32 +34,62 @@ func Username() string {
...
 
 func SecretKey() []byte {
 	return []byte("RandomizeThisPlease")
 }
 
-func NewToken() *Token {
-	token := jwt.New(jwt.SigningMethodHS256)
-
-	claims := token.Claims.(jwt.MapClaims)
+func CreateToken() *Token {
```
Comment (r0mant) [inline accounts/auth.go:46]: I like that everything token-related lives in its own module, but as a little code organization tip, I would make it into a some sort of service struct which can be configured and embedded into other structs if needed / implementation substituted. Something like:

```go
type Authenticator interface {
  IssueToken() (*Token, error)
  ValidateToken(Token) error
}

type jwtAuthenticator struct { } // implements Authenticator
```

File: accounts/config.go [hunk 14]
```diff
@@ -2,34 +2,43 @@ package accounts
...
 		Username: username,
 		Port:     "8000",
+		Token:    token,
 	}
 }
 
 // Client initialization
 func FindClientConfig() *ClientConfig {
-	file := FindFile(paths)
+	file, err := FindFile(paths)
```
Comment (r0mant) [inline accounts/config.go:37]: All this code here seems to be dealing with parsing client configuration file which adds a lot of complexity unnecessary to the challenge. Can the client just accept stuff it needs on the CLI?

File: cmd/schedule.go [hunk 15]
```diff
@@ -1,26 +1,188 @@
...
+var cli_cmd string
+
+var job_flags = map[string]bool{
+	"tail":   true,
+	"output": true,
+	"status": true,
+	"remove": true,
+}
+
+func ParseArgs(username string) *jobs.Request {
```
Comment (r0mant) [inline cmd/schedule.go:50]: Similar to my previous feedback, the code from here and below seems like implements a CLI args/flags parser which adds more complexity. Is it possible to use one of the standard ones instead?

File: comm/client.go [hunk 16]
```diff
@@ -1,86 +1,87 @@
...
 	return http.DefaultClient
 }
 
-func MakeRequest(config *accounts.ClientConfig, request *Request) string {
+func TokenArray(config *accounts.ClientConfig) []string {
+	return []string{config.Token.Signed}
+}
+
+// Make a request to the server
+func MakeRequest(config *accounts.ClientConfig, request *jobs.Request) *Response {
```
Comment (r0mant) [inline comm/client.go:27]: Instead of having separate unrelated methods for the different client-related things, I would consider collecting them into a single `Client` interface which would have methods like `Schedule`, `StreamLogs`, etc. Same on the server side.

File: comm/server.go [hunk 17]
```diff
@@ -99,26 +116,79 @@ func HandleSighup() {
...
+	router.HandleFunc("/tail", Tail)
+	router.HandleFunc("/output", Output)
+	router.HandleFunc("/status", Status)
+	router.HandleFunc("/remove", Remove)
 
 	return router
 }
+
+// TODO: This is pretty horrible...
+func WatchFileChange() {
```
Comment (r0mant) [inline comm/server.go:153]: Why is this needed?

File: jobs/active.go [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+	Offset int64
+}
+
+func NewActiveJob(command *Command) *ActiveJob {
+	return &ActiveJob{
+		Cmd: command,
+	}
+}
+
+func NewImmediateJob(account *accounts.Account, command *Command) *ActiveJob {
```
Comment (r0mant) [inline jobs/active.go:34]: What's Active vs Immediate job?

File: jobs/convert.go [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+type TimeScale int
+
+var (
+	SEC  TimeScale = 1
+	MIN  TimeScale = 2
+	HOUR TimeScale = 4
+	DAY  TimeScale = 8
+)
+
+var lookup = map[string]TimeScale{
```
Comment (r0mant) [inline jobs/convert.go:16]: Why is this needed? What's time scale?

File: jobs/scheduler.go [hunk 20]
```diff
@@ -0,0 +1,302 @@
...
+		return nil
+	}
+	command, ok := user.Commands[jobid]
+	if !ok {
+		return nil
+	}
+	return command
+}
+
+func (sched *Scheduled) ResetCommand(command *Command) {
```
Comment (r0mant) [inline jobs/scheduler.go:188]: What's reset command and why is it needed?

**Review by r0mant — COMMENTED** [review]
My overall feedback stays the same as for the first PR: I think that the scope of the project has become very broad. I would attempt to **drastically** simplify everything. For example, things like: the datastore, dynamically registering accounts, parsing configuration files, parsing CLI arguments, hot reloads - all of this is unnecessary and only adds a lot of extra complexity to the project. I would try to keep it very simple - keep only the essential stuff and make it the best possible production quality. cc @pierrebeaucamp @knisbet

File: accounts/accounts.go (hunk 1 above)
Comment (phomer) [reply accounts/accounts.go:56]: I kinda like prefacing it with Default, but isn't that a bit misleading if the struct is initialized from persistence? Ironically, it feels like I should call it OldAuthentication :-) (note: I don't really like calling the struct Authentication either, it's misnamed as well.)

File: comm/handlers.go (hunk 5 above)
Comment (phomer) [reply comm/handlers.go:276]: Way, way back, on a project, we did have a strict coding convention where we had to use any 'primitives' fully, before adding in new ones. So, if Sleep could be used, that would be preferable to Ticker. I did like that as a convention, in that it minimized the primitives and libraries. It did mean slightly longer code, but it was generally easier to understand. Often I've seen the opposite convention where people want to maximize libraries, technologies and function calls. If that happens, when the system gets large, it becomes really painful to keep running.

File: accounts/config.go (hunk 3 above)
Comment (phomer) [reply accounts/config.go:97]: Opps, I think the Glob should be on a Join with the path. Typo :-)

File: comm/handlers.go (hunk 4 above)
Comment (phomer) [reply comm/handlers.go:185]: Yea, it's pretty hacky. If I were working normally I would have tried to wire in something above all of this to deal with panics, validations, etc. It's very kludgy and repetitive right now.

File: accounts/auth.go (hunk 2 above)
Comment (phomer) [reply accounts/auth.go:61]: It's was just scaffolding temporarily placed there to find a calling issue with my Parse call (I think Marshal kept claiming it couldn't unpack the Claims structure or something. I was not calling with the right values). It should be removed.

File: accounts/accounts.go (hunk 7 above)
Comment (phomer) [reply accounts/accounts.go:103]: That's actually a really old coding standard issue. We had a strong preference for explicit 'symmetry' in the code. So if the code started with getting a resource or locking or some other symmetric operation, then it should end with them as well. As a convention, it allowed us to visually check that all of the resources, locks, etc. were cleaned up correctly. I do find it odd that Go went far away from that convention, the defer syntax always feels kinda noisy to me. So out of habit, I tend to type it in 'old style', then if I remember, I change it to Go conventions. I think I would have preferred them adding in critical section blocks, indented and bounded by braces.

File: comm/handlers.go (hunk 8 above)
Comment (phomer) [reply comm/handlers.go:34]: Scaffolding again. Should be deleted.

File: comm/handlers.go (hunk 5 above)
Comment (phomer) [reply comm/handlers.go:276]: I think you are right about that. I had a race between the process exiting and the file being synced, but I should have taken care of that below.

File: comm/server.go (hunk 10 above)
Comment (phomer) [reply comm/server.go:128]: I had figured since I should handle SIGCHLD for reaping anyways, that SIGINT for reloading would be fine. I was surprised to find out that was interfering with net/http, then again later with fsnotify. If I had time I would have dug into what that was happening (since I find it kinda scary).
